packages/api/src/derivedProperties/WithPropertyObjectSet.ts

```diff
@@ -0,0 +1,119 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { ValidAggregationKeys } from "../aggregate/AggregatableKeys.js";
+import type { WhereClause } from "../aggregate/WhereClause.js";
+import type {
+  ObjectOrInterfaceDefinition,
+  PropertyKeys,
+} from "../ontology/ObjectOrInterface.js";
+import type {
+  CompileTimeMetadata,
+  PropertyDef,
+} from "../ontology/ObjectTypeDefinition.js";
+import type { LinkedType, LinkNames } from "../util/LinkUtils.js";
+import type { WithPropertyDefinition } from "./WithPropertiesClause.js";
+
+export interface WithPropertyObjectSet<Q extends ObjectOrInterfaceDefinition>
+  extends
+    BaseWithPropertyObjectSet<Q>,
+    AggregatableWithPropertyObjectSet<Q>,
+    SingleLinkWithPropertyObjectSet<Q>
+{}
+
+export interface BaseWithPropertyObjectSet<
+  Q extends ObjectOrInterfaceDefinition,
+> extends FilterableDeriveObjectSet<Q> {
+  readonly pivotTo: <L extends LinkNames<Q>>(
+    type: L,
+  ) => NonNullable<CompileTimeMetadata<Q>["links"][L]["multiplicity"]> extends
+    false ? SingleLinkWithPropertyObjectSet<LinkedType<Q, L>>
+    : ManyLinkWithPropertyObjectSet<LinkedType<Q, L>>;
+}
+
+interface FilterableDeriveObjectSet<
+  Q extends ObjectOrInterfaceDefinition,
+> {
+  readonly where: (
+    clause: WhereClause<Q>,
+  ) => this;
+}
+
+type CollectAggregations = "collectSet" | "collectList";
+
+type BaseAggregations =
+  | "approximateDistinct"
+  | "exactDistinct"
+  | "approximatePercentile";
+
+type StringDeriveAggregateOption =
+  | BaseAggregations
+  | CollectAggregations;
+
+type NumericDeriveAggregateOption =
+  | "min"
+  | "max"
+  | "sum"
+  | "avg"
+  | BaseAggregations
+  | CollectAggregations;
+
+interface AggregatableWithPropertyObjectSet<
+  Q extends ObjectOrInterfaceDefinition,
+> extends FilterableDeriveObjectSet<Q> {
+  readonly aggregate: <
+    V extends ValidAggregationKeys<
+      Q,
+      StringDeriveAggregateOption,
+      NumericDeriveAggregateOption
+    >,
+  >(
+    aggregationSpecifier: V,
+    opts?: V extends `${any}:${infer P}`
+      ? P extends CollectAggregations ? { limit: number }
+      : P extends "approximatePercentile" ? { percentile: number }
+      : never
+      : never,
+  ) => WithPropertyDefinition<
+    V extends `${infer N}:${infer P}`
+      ? P extends CollectAggregations ? PropertyDef<
+          CompileTimeMetadata<Q>["properties"][N]["type"],
+          "nullable",
+          "array"
+        >
+      : P extends "approximateDistinct" | "exactDistinct" | "$count"
+        ? PropertyDef<"integer">
+      : PropertyDef<"double">
+      : V extends "$count" ? PropertyDef<"integer">
+      : never
+  >;
+}
+
+interface SingleLinkWithPropertyObjectSet<
```

Comment:
`BaseObjectSet` extends `SingleLinkWithPropertyObjectSet`. This makes the `selectProperty` method available only for chains of single links, which correlates with the backend. The `pivotTo` operation in `BaseObjectSet` enforces that if we pivot to a many link, we from then on use `ManyLinkWithPropertyObjectSet` which does not have `selectProperty` available and only returns further ManyLink object sets.